File: sdk/conformance/harness/python/conformance/diff_cli.py
"""diff-cli: reads parity scenarios from stdin, emits per-case
verdict JSON on stdout. Used by parity/run.sh to cross-check this
harness's diff against the Go + TypeScript equivalents.

    python -m conformance.diff_cli < parity/scenarios.json > verdicts-python.json
"""

from __future__ import annotations

import json
import sys

from . import _diff


def main() -> int:
    raw = sys.stdin.read()
    scenarios = json.loads(raw)
    verdicts = []
    for s in scenarios:
        name = s.get("name", "")
        want = bool(s.get("want_pass", False))
        try:
            _diff.diff(s.get("actual"), s.get("expected"), "$")
            passed = True
            err = ""
        except _diff.DiffError as exc:
            passed = False
            err = str(exc)
        verdicts.append(
            {
                "name": name,
                "passed": passed,
                "error": err,
                "want_pass": want,
                "agreed_with_want": passed == want,
            }
        )
    json.dump(verdicts, sys.stdout, indent=2)
    sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: sdk/conformance/harness/typescript/src/diff-cli.mjs
#!/usr/bin/env node
// diff-cli: reads parity scenarios from stdin, emits per-case verdict
// JSON on stdout. Used by parity/run.sh to cross-check this
// harness's diff against Go + Python.
//
//   node src/diff-cli.mjs < parity/scenarios.json > verdicts-typescript.json

import { diff, DiffError } from './diff.mjs';

async function readStdin() {
  const chunks = [];
  for await (const chunk of process.stdin) chunks.push(chunk);
  return Buffer.concat(chunks).toString('utf8');
}

async function main() {
  const raw = await readStdin();
  const scenarios = JSON.parse(raw);
  const verdicts = scenarios.map((s) => {
    const want = Boolean(s.want_pass);
    let passed = true;
    let err = '';
    try {
      diff(s.actual, s.expected, '$');
    } catch (e) {
      if (!(e instanceof DiffError)) throw e;
      passed = false;
      err = e.message;
    }
    return {
      name: s.name,
      passed,
      error: err,
      want_pass: want,
      agreed_with_want: passed === want,
    };
  });
  process.stdout.write(JSON.stringify(verdicts, null, 2) + '\n');
}

main().catch((err) => {
  process.stderr.write(`diff-cli: ${err.stack || err.message}\n`);
  process.exit(2);
});
